workers/unity/Packages/com.improbable.gdk.simlauncher/DeploymentLauncher.cs

```diff
@@ -0,0 +1,258 @@
+using System;
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.IO;
+using System.Text;
+using System.Threading.Tasks;
+using Fps;
+using Improbable.Gdk.Core;
+using Improbable.Gdk.Tools;
+using UnityEditor;
+using UnityEngine;
+using Debug = UnityEngine.Debug;
+
+namespace Improbable.Gdk.SimLauncher
+{
+    public class DeploymentLauncher
+    {
+        private static readonly string DotNetProjectPath =
+            Path.GetFullPath(Path.Combine(Tools.Common.GetPackagePath("com.improbable.gdk.simlauncher"), ".DeploymentLauncher/DeploymentLauncher.csproj"));
+
+        private static readonly string ProjectRootPath =
+            Path.Combine(Application.dataPath, "../../../");
+
+        private const string SimLauncherMenuItem = "SpatialOS/Launch Deployment with simulated players";
+
+        private const int SimLauncherPriority = 51;
+
+        private const string ProjectName = "unity_gdk";
+
+        private class LaunchDeploymentEditorWindow : EditorWindow
+        {
+            private string assemblyName;
+            private string deploymentName;
+            private string snapshotPath = "snapshots/cloud.snapshot";
+            private string mainLaunchJson = "cloud_launch_small.json";
+            private string simWorkerLaunchJson = "cloud_launch_small_sim_workers.json";
+
+            private bool deploymentRunning;
+
+            private Task runningLaunchTask;
+            private bool simWorkerDeploymentEnabled;
+
+            void OnGUI()
+            {
+                assemblyName = EditorGUILayout.TextField("Assembly Name", assemblyName);
+                deploymentName = EditorGUILayout.TextField("Deployment Name", deploymentName);
+                snapshotPath = EditorGUILayout.TextField("Snapshot Path", snapshotPath);
+                if (runningLaunchTask != null)
+                {
+                    if (GUILayout.Button("Cancel Launch"))
+                    {
+                        OnClickShutdownDeployment();
+                    }
+
+                    if (runningLaunchTask.Wait(0))
+                    {
+                        deploymentRunning = true;
```

Comment:
On second thought, I recommend for these kind of things (UI state altering logic) to be within an Update function of your EditorWindow. Then call `Repaint` when you modify anything that would alter the UI state.

Here's my thoughts that lead to this:

Depends on which imgui event is current one. Depending on which event it is it may cause problems or a stale GUI.

Event order is roughly:
- something happens that triggers a GUI dispatch
- frame starts
- layout, positioning
- repaint
- input stuff (mouse, keyboard)
- frame ends

If this condition becomes true within a repaint or as part of the input events, then it will result in an error in the next event because it will result in a different number of gui elements within the layout group.

If this condition becomes true at the last event by chance, it's possible that it will not trigger the next gui dispatch until "something in the editor updates", which could be very soon or much later by chance. I think it's recommended to alter "how many things will be drawn into this window" only in the layout phase. Or outside of the OnGUI if possible. E.g. within an `Update` event.

Test script:

```cs
    public class TestWindow : EditorWindow
    {
        public int counter;
        public bool repaint;

        public void Update()
        {
            ++counter;

            if (repaint)
            {
                Repaint();
            }
        }

        [MenuItem("SpatialOS/Test Window")]
        public static void StartWindow()
        {
            GetWindow<TestWindow>().Show();
        }


        public void OnGUI()
        {
            EditorGUILayout.LabelField($"Count: {counter}");

            repaint = EditorGUILayout.Toggle("Repaint", repaint);
        }
    }
```

https://docs.unity3d.com/Manual/ExecutionOrder.html